test/functional/wallet_reorgsrestore.py

```diff
@@ -31,6 +31,41 @@ def set_test_params(self):
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
 
+    def test_reorg_handling_during_unclean_shutdown(self):
+        self.log.info("Test crash due to a duplicate block disconnection event during an unclean shutdown")
+        node = self.nodes[0]
+        # Receive coinbase reward on a new wallet
+        node.createwallet(wallet_name="reorg_crash", load_on_startup=True)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        self.generatetoaddress(node, 1, wallet.getnewaddress(), sync_fun=self.no_op)
+
+        # Restart to ensure node and wallet are flushed
+        self.restart_node(0)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        assert_equal(wallet.getwalletinfo()['immature_balance'], 25)
+
+        # Disconnect tip
+        tip = wallet.getbestblockhash()
+        wallet.invalidateblock(tip)
+        wallet.syncwithvalidationinterfacequeue()
+
+        # Tip was disconnected, ensure coinbase has been abandoned
+        assert_equal(wallet.getwalletinfo()['immature_balance'], 0)
+        coinbase_tx_id = wallet.getblock(tip, verbose=1)["tx"][0]
+        assert_equal(wallet.gettransaction(coinbase_tx_id)['details'][0]['abandoned'], True)
+
```

Comment:
Interesting that the best block hash at the stage is not the `tip` because of invalidation but not after unclean-stop-start.

```
assert_equal(node.getbestblockhash() == tip, False)
```